### PR TITLE
Task4

### PR DESCRIPTION
**Pull request by pakhuta**

### CODE REVIEW

File: src/routes/postProducts.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import Storage from '../services/storage';
+
+export default function getProducts(req, res) {
+    res.send(Storage.add('products', req.body));
```
Comment (boccob) [inline src/routes/postProducts.js]: Hmm, doesn't work for me for either x-www-form-urlencoded or application/json request.

File: src/routes/getProduct.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import Storage from '../services/storage';
+
+export default function getProduct(req, res) {
+    let product = Storage.get('products', [{ name: 'id', value: req.params.id }]) || {};
```
Comment (boccob) [inline src/routes/getProduct.js]: What if there is no such product?

File: src/services/storage/index.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import fs from 'fs';
```
Comment (boccob) [inline src/services/storage/index.js]: Too complex for my taste. Don't know why you need two classes here.

File: src/app.js [hunk 4]
```diff
@@ -1,15 +1,30 @@
...
+import queryParser from './middlewares/queryParser';
+import ResponseUtils from './utils/responseUtils';
+import routes from './routes';
+
+const app = express();
+
+app.use(cookiesParser);
+app.use(queryParser);
+app.use(express.json());
+app.use((err, req, res, next) => {
```
[the remark this answers is not in the document]
Comment (pakhuta) [reply src/app.js]: What do you mean? I use it for handle errors http://expressjs.com/en/guide/error-handling.html. Or you mean that it must be the last of all `app.use()`

File: src/http-servers/echo-server.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import http from 'http';
```
[the remark this answers is not in the document]
Comment (pakhuta) [reply src/http-servers/echo-server.js]: Do you mean run them on different ports?

File: src/http-servers/html-server.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import File from '../utils/file';
+import ResponseUtils from '../utils/responseUtils';
+
+const server = http.createServer(requestHandler);
+
+server.listen(config.server.port, () => {
+    Output.write(util.format('Web server has started at port: %s', config.server.port));
+});
+
+function requestHandler(req, res) {
```
[the remark this answers is not in the document]
Comment (pakhuta) [reply src/http-servers/html-server.js]: Fixed in a task 5

File: src/middlewares/queryParser.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import url from 'url';
+
+export default function queryParser(req, res, next) {
+    try {
+        req.parsedQuery = url.parse(req.url, true).query;
```
[the remark this answers is not in the document]
Comment (pakhuta) [reply src/middlewares/queryParser.js]: Fixed in a task 5

File: src/routes/getProduct.js (hunk 2 above)
Comment (pakhuta) [reply src/routes/getProduct.js]: If will not be found product with id === req.params.id, will be responded with an empty array.

Comment (pakhuta) [reply src/routes/getProduct.js]: I don't quite understand the question, but I storage `reviews` in a memory and in a JSON as an array:
https://github.com/pakhuta/siarhei.pakhuta.nodejs/blob/5ece536df72039f47b1f31463847e6749f877eeb/src/models/product.js#L5
https://github.com/pakhuta/siarhei.pakhuta.nodejs/blob/5ece536df72039f47b1f31463847e6749f877eeb/data/products.json#L5

File: src/routes/getProductReviews.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+import Storage from '../services/storage';
+
+export default function getProductReviews(req, res) {
+    let { reviews = [] } = Storage.get('products', [{ name: 'id', value: req.params.id }]) || {};
```
[the remark this answers is not in the document]
Comment (pakhuta) [reply src/routes/getProductReviews.js]: I don't quite understand the question, but I storage `reviews` in a memory and in a JSON as an array:
https://github.com/pakhuta/siarhei.pakhuta.nodejs/blob/5ece536df72039f47b1f31463847e6749f877eeb/src/models/product.js#L5
https://github.com/pakhuta/siarhei.pakhuta.nodejs/blob/5ece536df72039f47b1f31463847e6749f877eeb/data/products.json#L5

File: src/routes/postProducts.js (hunk 1 above)
Comment (pakhuta) [reply src/routes/postProducts.js]: Must mork, please check format for application/json: https://i.imgur.com/RurCQIP.png

File: src/services/storage/index.js (hunk 3 above)
Comment (pakhuta) [reply src/services/storage/index.js]: Fixed in task 5

File: src/app.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (boccob) [reply src/app.js]: Sorry, I forgot about error middlewares. But yes, it'd be better to place it at the end of the chain.

File: src/http-servers/echo-server.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (boccob) [reply src/http-servers/echo-server.js]: Possibility to run them with `npm run`.

File: src/routes/getProduct.js (hunk 2 above)
Comment (boccob) [reply src/routes/getProduct.js]: Better to respond with 404.

File: src/routes/getProductReviews.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (boccob) [reply src/routes/getProductReviews.js]: `Storage.get` is not very clear. `Storage.get('products')` returns an array, but if you add a filter, it returns array item.
I'd prefer to separate concerns. `Storage.get('products').filter()`. For me, it looks more obvious. And I like obvious code.

**Pull request merged**